### ISSUE TITLE
Make Horovod run with CPUs and non-zero rank

### ISSUE THREAD

**Issue by tvovalentin** [horovod/horovod#1237]
**Environment:**
1. Framework: Keras
2. Framework version: 2.2.4
3. Horovod version: 0.16.4
4. MPI version: openmpi-bin 1.10.2-8ubuntu1
5. CUDA version:  -
6. NCCL version:  -
7. Python version: 3.5.2
8. OS and version: Ubuntu 16.04.5 LTS x86_64
9. GCC version: 5.4.0 20160609

**Checklist:**
1. Did you search issues to find if somebody asked this question before? Yes.
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)? Yes.
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? N/A.
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? Yes.

**Your question:**
I am trying to train a Keras model on a cluster with 3 nodes, so that my training time decreases, without affecting my accuracy. Because I am new to the whole Keras-Horovod-Spark integration, I am trying each part individually. So first, I want to run a simple script on my 3 nodes and learn how it works. I have defined the passwordless ssh access from the VMs in the cluster, created a NFS folder, and I have a simple script inside:

```
import os
import socket

from keras import backend as K
import horovod.tensorflow as hvd
import tensorflow as tf

print("hey", socket.gethostname(), ":", os.getcwd())

hvd.init()
print('Hello, rank = %d, local_rank = %d, size = %d, local_size = %d' % (hvd.rank(), hvd.local_rank(), hvd.size(), hvd.local_size()))
```

Because my goal is running that on a cluster of CPUs only (we don't have GPUs at all), preferably with Spark, I am issuing this command from one of my nodes (the master, called bd1):

```
horovodrun -np 2 -H bd1,bd2 python horovod_test_hey.py
```

The problem is that my output reveals a rank of 0 for both - how can I fix that so it works with CPUs as well? In the case I try to run my NN model, it fails because two processes try to write to the same file (a checklist after each epoch).

Here's the output: 

```
[1,0]<stdout>:hey bd1 : /data/JupyterNotebooks/ml/horovod/cloud
[1,1]<stdout>:hey bd2 : /data/JupyterNotebooks/ml/horovod/cloud
[1,0]<stderr>:2019-07-19 16:40:25.430104: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
[1,1]<stderr>:2019-07-19 16:40:25.431019: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
[1,0]<stderr>:2019-07-19 16:40:25.436883: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 1696070000 Hz
[1,1]<stderr>:2019-07-19 16:40:25.437092: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 1696070000 Hz
[1,0]<stderr>:2019-07-19 16:40:25.437350: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x4274340 executing computations on platform Host. Devices:
[1,0]<stderr>:2019-07-19 16:40:25.437377: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
[1,1]<stderr>:2019-07-19 16:40:25.437555: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x58a2860 executing computations on platform Host. Devices:
[1,1]<stderr>:2019-07-19 16:40:25.437588: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
[1,0]<stdout>:Hello, rank = 0, local_rank = 0, size = 1, local_size = 1
[1,0]<stderr>:Using TensorFlow backend.
[1,1]<stdout>:Hello, rank = 0, local_rank = 0, size = 1, local_size = 1
[1,1]<stderr>:Using TensorFlow backend.
```

Thank you kindly!

**Comment by alsrgv** [comment]
@tvovalentin, most likely you have different versions of MPI in your system. Also, OpenMPI 1.10 is super old. Could you uninstall all MPIs that you have in your system and install Open MPI 4.0.1?

**Comment by tvovalentin** [comment]
Sure @alsrgv, I have done that, but the problem persists: my rank is always 0, no matter how many instances I run. I am running it on two VMs (as root on both), by the way.

I have tried all the following (from bd1):

```
# horovodrun -np 4 -H bd1:2,bd2:2 python horovod_test_hey.py
# horovodrun -np 2 -H bd1,bd2 python horovod_test_hey.py
# horovodrun -np 2 -H bd1:2 python horovod_test_hey.py
# horovodrun -np 2 -H bd2:2 python horovod_test_hey.py
# horovodrun -np 2 -H localhost:2 python horovod_test_hey.py
# horovodrun -np 4 -H localhost:2,localhost:2 python horovod_test_hey.py 
```

**Comment by alsrgv** [comment]
@tvovalentin, what's the output of `mpirun --version`?  Did you reinstall Horovod after installing Open MPI 4.0.1?

**Comment by tvovalentin** [comment]
Hi @alsrgv !

I managed to get it working now (with version 4.0.1). I had tried to reinstall Horovod, but it wasn't reinstalled properly. I fixed it by forcing the reinstall: `pip install horovod --force-reinstall --no-cache-dir`

Thank you very much for the help! Cheers.

**Comment by LANRRI** [comment]
@tvovalentin hello, i'm new to horovod, can we submit our program to spark cluster using horovod?

**Comment by LANRRI** [comment]
@tvovalentin thanks. i've got the answer

**Comment by pl-lee** [comment]
@tvovalentin Hi, have you solved this problem? My Open MPI is 4.0.0, and my output is same to you. could anybody tell me the reason? Thanks!

**Comment by cosmichut** [comment]
> Hi @alsrgv !
> 
> I managed to get it working now (with version 4.0.1). I had tried to reinstall Horovod, but it wasn't reinstalled properly. I fixed it by forcing the reinstall: `pip install horovod --force-reinstall --no-cache-dir`
> 
> Thank you very much for the help! Cheers.

It helps exactly to solve the same problem..thanks a lot.
